### PR TITLE
tsp, support sample codegen for grouping and flattening

### PR DESCRIPTION
**Pull request by XiaofeiCao**
issue: https://github.com/Azure/autorest.java/issues/2201

Grouping and flattening implementation design by weidong: https://github.com/Azure/autorest.java/blob/main/docs/developer/parameter-transformation.md
#### 1. Grouping without flattening
https://github.com/Azure/autorest.java/blob/88842c5eea5c460742d6d057345f78c0cf4d0942/typespec-tests/tsp/special-headers.tsp#L52-L73

ClientModel: 
```yaml
outParameter: ifMatch
parameterMappings:
- inputParameter: requestConditions
  inputParameterProperty: ifMatch
```

#### 2. Flattening without grouping
https://github.com/Azure/autorest.java/blob/657243781a29ae8ff276e7514cc22d5960ba7e8b/typespec-tests/tsp/flatten.tsp#L52-L54

ClientModel: 
```yaml
outParameter: request
parameterMappings:
- inputParameter: user
  outputParameterPropertyName: user
- inputParameter: input
  outputParameterPropertyName: input
```

#### 3. Flattening then grouping
https://github.com/Azure/autorest.java/blob/657243781a29ae8ff276e7514cc22d5960ba7e8b/typespec-tests/tsp/flatten.tsp#L56-L58

```yaml
- outParameter: id
  parameterMappings:
    - inputParameter: options
      inputParameterProperty: id
- outParameter: filter
  parameterMappings:
    - inputParameter: options
      inputParameterProperty: filter
- outParameter: request # flattened(real) parameter
  parameterMappings:
    - inputParameter: options # grouping parameter
      inputParameterProperty: user # grouped(real) parameter
      outputParameterPropertyName: user
      outputParameterProperty: user
    - inputParameter: options # grouping parameter
      inputParameterProperty: input # grouped(real) parameter
      outputParameterPropertyName: input
      outputParameterProperty: input
```

Major code change in file `ClientMethodExampleWriter` https://github.com/Azure/autorest.java/pull/2236/files#diff-beed9a388e5f2f63c845e612d9203ba702a646bb6ee183dd41caf741eb1aa49e

Others are bug fixes or code refactors.

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java [hunk 1]
```diff
@@ -333,12 +333,8 @@ private void addConvenienceExamples(Client.Builder builder, List<AsyncSyncClient
...
                         .isExampleIncluded(clientMethod, convenienceMethod))
                     .max((clientMethod1, clientMethod2) -> {
-                        int m1ParameterCount = clientMethod1.getOnlyRequiredParameters()
-                            ? clientMethod1.getMethodRequiredParameters().size()
-                            : clientMethod1.getMethodParameters().size();
-                        int m2ParameterCount = clientMethod2.getOnlyRequiredParameters()
-                            ? clientMethod2.getMethodRequiredParameters().size()
-                            : clientMethod2.getMethodParameters().size();
+                        int m1ParameterCount = clientMethod1.getMethodInputParameters().size();
+                        int m2ParameterCount = clientMethod2.getMethodInputParameters().size();
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientMapper.java:337]: simplify code, logic is the same

File: javagen/src/main/java/com/azure/autorest/template/example/ModelExampleWriter.java [hunk 2]
```diff
@@ -269,7 +269,7 @@ public String accept(ExampleNode node) {
                     List<ClientModelProperty> requiredParentProperties = ClientModelUtil.getRequiredWritableParentProperties(model);
                     List<ClientModelProperty> requiredProperties = model.getProperties().stream()
                             .filter(ClientModelProperty::isRequired)
-                            .filter(property -> !property.isConstant())
+                            .filter(property -> !property.isConstant() && !property.isReadOnly())
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/template/example/ModelExampleWriter.java:272]: Bug fix, constructor args with required read-only parameter.

File: javagen/src/main/java/com/azure/autorest/util/ModelExampleUtil.java [hunk 3]
```diff
@@ -240,16 +240,42 @@ public static Object getChildObjectValue(List<String> jsonPropertyNames, Object
...
-     * @return
+     * @return example node
      */
     public static ExampleNode parseNodeFromParameter(ProxyMethodExample example, MethodParameter methodParameter) {
         String serializedName = methodParameter.getSerializedName();
         if (serializedName == null && methodParameter.getProxyMethodParameter().getRequestParameterLocation() == RequestParameterLocation.BODY) {
             serializedName = methodParameter.getProxyMethodParameter().getName();
         }
 
+        Object exampleValue = getParameterExampleValue(example, serializedName, methodParameter.getProxyMethodParameter().getRequestParameterLocation());
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/util/ModelExampleUtil.java:251]: refactor, extract `getParameterExampleValue` method for reuse

File: typespec-tests/Setup.ps1 [hunk 4]
```diff
@@ -1,3 +1,12 @@
+param (
+    # re-build autorest.java
+    [switch] $RebuildJar = $false
+)
+
+if ($RebuildJar) {
+    mvn clean install package -f ../pom.xml -Plocal -Ptsp -DskipTests
+}
+
```
Comment (XiaofeiCao) [inline typespec-tests/Setup.ps1:9]: For local run.

E.g. `pwsh Setup.ps1 -RebuildJar` will rebuild autorest.java for typespec-java.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java [hunk 5]
```diff
@@ -57,6 +66,106 @@ public ClientMethodExampleWriter(ClientMethod method, String clientVarName, Prox
...
+            return ModelExampleUtil.parseNode(type, wireType, exampleValue);
+        } else if (isFlattenParameter(convenienceMethod, methodParameter)) {
+            // flatten, no grouping
+            String realParameterName = convenienceMethod.getMethodTransformationDetails().iterator().next().getOutParameter().getName();
+            Map<String, Object> realParameterValue = (Map<String, Object>) proxyMethodExample.getParameters().get(realParameterName).getObjectValue();
+
+            IType type = methodParameter.getClientMethodParameter().getClientType();
+            IType wireType = methodParameter.getClientMethodParameter().getWireType();
+
+            Object flatteningParameterValue = null;
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:112]: `flattenedParameterValue`? or maybe `methodParameterValue`?

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java [hunk 6]
```diff
@@ -57,6 +66,106 @@ public ClientMethodExampleWriter(ClientMethod method, String clientVarName, Prox
...
+            if (realParameterValue != null) {
+                flatteningParameterValue = realParameterValue
+                        .entrySet()
+                        .stream()
+                        .filter(entry -> Objects.equals(
+                                CodeNamer.getEscapedReservedClientMethodParameterName(entry.getKey()),
+                                methodParameter.getClientMethodParameter().getName()))
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:119]: I am not sure about this part. Should the comparison be based on "serialized name"?

And the logic here is different from the "flatten" process in grouping?

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java [hunk 7]
```diff
@@ -57,6 +66,106 @@ public ClientMethodExampleWriter(ClientMethod method, String clientVarName, Prox
...
+                        serializedParameterName = detail.getOutParameter().getName();
+                        Map<String, Object> flattenedParameterValue = (Map<String, Object>) ModelExampleUtil.getParameterExampleValue(proxyMethodExample, serializedParameterName, parameter.getRequestParameterLocation());
+                        if (flattenedParameterValue != null) {
+                            exampleValue.putAll(flattenedParameterValue);
+                        }
+                    } else {
+                        exampleValue.put(serializedParameterName,
+                                ModelExampleUtil.getParameterExampleValue(
+                                        proxyMethodExample, serializedParameterName, parameter.getRequestParameterLocation()));
+                    }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:98]: I assume the correctness of the logic depends on that "serialized name" of real parameter would be same as "serialized name" of the group property. This equal is is implicit (as "serialized name" of the group property won't have effect on wire, so strictly speaking it doesn't have to be that value).

If so, recommend to note this in comment.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java [hunk 8]
```diff
@@ -57,6 +66,106 @@ public ClientMethodExampleWriter(ClientMethod method, String clientVarName, Prox
...
+        return details.stream().allMatch(
+                detail ->
+                        !CoreUtils.isNullOrEmpty(detail.getParameterMappings())
+                                && detail.getOutParameter() != null
+                                &&
+                                // same name
+                                Objects.equals(
+                                        CodeNamer.getEscapedReservedClientMethodParameterName(
+                                                detail.getParameterMappings().iterator().next().getInputParameter().getName()),
+                                        methodParameter.getClientMethodParameter().getName())
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:144]: again, I am not sure about this.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java (hunk 7 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:98]: Exactly. The group property's name is actually from real parameter's `serializedName` in emitter. Will add comment about this.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java (hunk 6 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java]: Changed the check the outputParameterProperty's serializedName.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java (hunk 8 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java]: Changed to check inputParameter's name against the client method parameter's name.

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java [hunk 9]
```diff
@@ -57,6 +66,125 @@ public ClientMethodExampleWriter(ClientMethod method, String clientVarName, Prox
...
+                    }
+                }
+            }
+            IType type = methodParameter.getClientMethodParameter().getClientType();
+            IType wireType = methodParameter.getClientMethodParameter().getWireType();
+            return ModelExampleUtil.parseNode(type, wireType, exampleValue);
+        } else if (isFlattenParameter(convenienceMethod, methodParameter)) {
+            // flatten, no grouping
+            String realParameterName = convenienceMethod.getMethodTransformationDetails().iterator().next().getOutParameter().getName();
+            Map<String, Object> realParameterValue = (Map<String, Object>) proxyMethodExample.getParameters().get(realParameterName).getObjectValue();
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:115]: should it also call `findFlattenedBodyParameterValue`?

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java (hunk 9 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java:115]: Yes it should. Thanks!

Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/example/ClientMethodExampleWriter.java]: Changed.

**Pull request merged**